Fee estimation: avoid serving stale fee estimate

ismaelsadeeq opened the pull request:
Fixes #27555

The issue arises when an old `fee_estimates.dat` file is sometimes read during initialization.
Or after an unclean shutdown, the latest fee estimates are not flushed to `fee_estimates.dat`.
If the fee estimates in the old file are old, they can cause transactions to become stuck in the mempool. 
This  PR ensures that nodes do not use stale estimates from the old file during initialization. If  `fee_estimates.dat` 
has not been updated for 60 hours or more, it is considered stale and will not be read during initialization. To avoid 
having old estimates, the `fee_estimates.dat` file will be flushed periodically every hour. As mentioned #27555

> "The immediate improvement would be to store fee estimates to disk once an hour or so to reduce the chance of having an old file. From there, this case could probably be detected, and refuse to serve estimates until we sync."

In addition, I will follow-up PR to persist the `mempoolminfee` across restarts.

fanquake commented:
cc @instagibbs @TheBlueMatt

glozow commented on src/policy/fees.cpp:
```diff
@@ -12,6 +12,7 @@
 #include <policy/feerate.h>
 #include <primitives/transaction.h>
 #include <random.h>
+#include <scheduler.h>
```
I think it's a bit strange for policy/fees to have a dependency on scheduler. Maybe register `FlushFeeEstimates()` with the scheduler within `AppInitMain()` instead?

ismaelsadeeq replied:
Sure :+1: , fixed.

maflcko reviewed (approved):
lgtm. Left some nits

instagibbs commented on src/policy/fees.h:
```diff
@@ -14,12 +14,17 @@
 
+
+// How often to flush fee estimates to fee_estimates.dat.
+static constexpr std::chrono::milliseconds FLUSH_INTERVAL{std::chrono::hours{1}};
```
s/FLUSH_INTERVAL/FEE_FLUSH_INTERVAL/ ?

ismaelsadeeq replied:
FEE_FLUSH_INTERVAL is better.
I think.

ismaelsadeeq commented on src/policy/fees.cpp:
```diff
@@ -903,11 +904,15 @@ CFeeRate CBlockPolicyEstimator::estimateSmartFee(int confTarget, FeeCalculation
     if (est_file.IsNull() || !Write(est_file)) {
         LogPrintf("Failed to write fee estimates to %s. Continue anyway.\n", fs::PathToString(m_estimation_filepath));
     }
+    LogPrintf("Flushed fee estimates to fee_estimates.dat.\n");
```
Yes, fixed thank you.

ismaelsadeeq commented on test/functional/feature_fee_estimation.py:
```diff
@@ -290,6 +291,37 @@ def sanity_check_rbf_estimates(self, utxos):
+    def test_estimate_dat_is_flushed_periodically(self):
+            # Check if the string "Flushed fee estimates to fee_estimates.dat." is present in the debug log file.
+            # If it is, it indicates that fee estimates have been successfully flushed to disk.
+            with self.nodes[0].assert_debug_log(expected_msgs=["Flushed fee estimates to fee_estimates.dat."], timeout=1):
```
Yes thanks for pointing this out, I added the check.
Thank you.

ismaelsadeeq commented on src/policy/fees.h:
```diff
@@ -186,6 +186,10 @@ class CBlockPolicyEstimator
     const fs::path m_estimation_filepath;
+
+    /** Fee estimate.dat must not be more than 12 hours old. */
+    static constexpr std::chrono::hours MAX_FILE_AGE{12};
```
Yeah, to enhance flexibility, I can create a follow-up PR to introduce a configuration variable for MAX_FILE_AGE,  This way users can specify the maximum age of estimates they do not want to accept?

ismaelsadeeq replied:
I was thinking that introducing flexibility would enable users to define their preferred threshold for considering fee estimates as stale. As @instagibbs suggested extending the default 12 hours further. what is your view on it?

willcl-ark replied:
Another use case could be so that testers can purposefully load (very) old fee estimate files to populate `estimatesmartfee`? https://github.com/bitcoin/bitcoin/issues/27415#issuecomment-1531711091

glozow replied:
Not sure if we should load old (i.e. potentially bad) data as it can skew the fee estimates both at startup and after more data has been collected. For testing, they could mock the time right? Just worried about adding a config option that's a bit footgunny.

ismaelsadeeq replied:
Yes, I agree with you. stale estimates should not be read. I have updated the MAX_FILE_AGE to 60 hours, which 2.5 days, as @instagibbs suggested that longer would be okay?

willcl-ark replied:
Can you explain the rationale behind the 60 hours chosen? Also perhaps you can update OP as it still mentions 12 hours :)

pinheadmz commented on src/policy/fees.h:
```diff
@@ -14,12 +14,17 @@
 
+
+// How often to flush fee estimates to fee_estimates.dat.
+static constexpr std::chrono::milliseconds FEE_FLUSH_INTERVAL{std::chrono::hours{1}};
```
I don't actually know for sure but couldn't this just be ...?

```cpp
static constexpr std::chrono::hours FEE_FLUSH_INTERVAL{1};
```

ismaelsadeeq replied:
Still a nub on cpp,
 `constexpr auto FEE_FLUSH_INTERVAL{1h};`  does not compile
I added instead `static constexpr std::chrono::hours FEE_FLUSH_INTERVAL{1};`

pinheadmz commented on test/functional/feature_fee_estimation.py:
```diff
@@ -290,6 +291,45 @@ def sanity_check_rbf_estimates(self, utxos):
+        fee_rate = self.nodes[0].estimatesmartfee(1)["feerate"]
+
+        # Restart node to ensure fee_estimate.dat file is read
+        self.stop_node(0)
+        self.start_node(0)
```
I think you can use `self.restart_node(0)` here

ismaelsadeeq replied:
Thanks fixed

pinheadmz reviewed (commented):
concept ACK

This should solve the stale fee issue, the test is clever and effective. I ensured the test fails without the patch and also checked the test failed when the mtime was set < 12 hours. Reviewed code and left some notes / questions

ismaelsadeeq commented on src/policy/fees.cpp:
```diff
@@ -903,11 +904,16 @@ CFeeRate CBlockPolicyEstimator::estimateSmartFee(int confTarget, FeeCalculation
     AutoFile est_file{fsbridge::fopen(m_estimation_filepath, "wb")};
     if (est_file.IsNull() || !Write(est_file)) {
         LogPrintf("Failed to write fee estimates to %s. Continue anyway.\n", fs::PathToString(m_estimation_filepath));
+        return;
```
Thanks for your review, I have fixed it.

ismaelsadeeq replied:
>Mine as well use this constant here for the filename?

:thinking: it is an anon namespace, not defined in policy/fees_args.h, 
I think it is okay as is,  or else

policy/fees_args.h
 ```
const char* GetFeeEstimatesFilename();
```
policy/fees.args.cpp
```
const char* GetFeeEstimatesFilename() {
    return FEE_ESTIMATES_FILENAME;
}
```
policy/fees.cpp
```
LogPrintf("Flushed fee estimates to %s.\n", GetFeeEstimatesFilename());
```
but if it's possible another way I will be glad to learn how.

pinheadmz replied:
Ok good point I see, well maybe just then...


```diff
diff --git a/src/policy/fees.cpp b/src/policy/fees.cpp
index 085f938f80..adb0ebc383 100644
--- a/src/policy/fees.cpp
+++ b/src/policy/fees.cpp
@@ -925,7 +925,7 @@ void CBlockPolicyEstimator::FlushFeeEstimates() {
     if (est_file.IsNull() || !Write(est_file)) {
         LogPrintf("Failed to write fee estimates to %s. Continue anyway.\n", fs::PathToString(m_estimation_filepath));
     } else {
-        LogPrintf("Flushed fee estimates to fee_estimates.dat.\n");
+        LogPrintf("Flushed fee estimates to %s.", fs::PathToString(m_estimation_filepath));
     }
 }
 

```

glozow commented on src/policy/fees.cpp:
```diff
@@ -546,9 +546,22 @@ CBlockPolicyEstimator::CBlockPolicyEstimator(const fs::path& estimation_filepath
+
+    // Whenever the fee estimation file is not present return early
+    if (est_file.IsNull()) {
+        LogPrintf("Failed to read fee estimates from %s. Continue anyway.\n", fs::PathToString(m_estimation_filepath));
```
in accf995cf095d7ff3d10c0403e70bdbc1273e852

Seems like the error message here should be something like "there is no fee estimates file" instead of "failed to read" ?

ismaelsadeeq replied:
Thanks fixed

glozow reviewed (commented):
Approach ACK

instagibbs commented on test/functional/feature_fee_estimation.py:
```diff
@@ -290,6 +291,44 @@ def sanity_check_rbf_estimates(self, utxos):
+                self.nodes[0].mockscheduler(hour)
+
+            # Verify that fee estimates were flushed and fee_estimates.dat file is created
+            assert_equal(os.path.isfile(fee_dat), True)
```
might be good to test that it does it *every hour*, and that the output files should be matching provided we don't see blocks in the interim, also, check that what is written on node shutdown also matches?

That way we can have higher assurance on the timing of the flushing and the contents matching what we already do.

ismaelsadeeq replied:
Sure, thank you, I updated the test.

instagibbs reviewed (commented):
Looking good, just a couple of test suggestions.

Also wondering if we should include a hidden argument to allow older files for testing reasons, re: https://github.com/bitcoin/bitcoin/issues/27415 

I've used it for systems I've built for integration testing, and not having to touch a file during testing to refresh it would be a simplification.

> For testing, they could mock the time right?

Depends on your test setup, really. It might interfere during integration tests which require a realistic clock?

glozow commented:
> Also wondering if we should include a hidden argument to allow older files for testing reasons, re: #27415
> 
> I've used it for systems I've built for integration testing, and not having to touch a file during testing to refresh it would be a simplification.
> 
> > For testing, they could mock the time right?
> 
> Depends on your test setup, really. It might interfere during integration tests which require a realistic clock?

Makes sense, seems fine as a regtest-only option for testing.

ismaelsadeeq commented on test/functional/feature_fee_estimation.py:
```diff
@@ -290,6 +291,44 @@ def sanity_check_rbf_estimates(self, utxos):
+            fee_dat = self.nodes[0].chain_path / "fee_estimates.dat"
+
+            # Verify that fee_estimates.dat does not exist
+            assert_equal(os.path.isfile(fee_dat), False)
```
Thank you, I added the delete attempt

ismaelsadeeq commented:
> Also wondering if we should include a hidden argument to allow older files for testing reasons, re: #27415
> I've used it for systems I've built for integration testing, and not having to touch a file during testing to refresh it would be a simplification.
> 
> > For testing, they could mock the time right?
> 
> Depends on your test setup, really. It might interfere during integration tests which require a realistic clock?
>
> Makes sense, seems fine as a regtest-only option for testing.

Sure I will follow up with allowing reading stale estimates but only on regtest for testing purposes.

willcl-ark commented on test/functional/feature_fee_estimation.py:
```diff
@@ -290,6 +291,77 @@ def sanity_check_rbf_estimates(self, utxos):
+        # Stop the node and backdate the fee_estimates.dat file for 61 hours
+        self.stop_node(0)
+        hour = 60 * 60
+        last_modified_time = time.time() - 61 * hour
```
I know we only use it once in the test, but if you re-touch this file it could be nice to use a named variable for the 60 hours, to make it clear where it comes from? (e.g. `MAX_FILE_AGE` would shadow the name used in _fees.h_)

ismaelsadeeq replied:
I agree it's better than the magic number, I have updated it.

willcl-ark reviewed (approved):
ACK 451ab8b513

Left a few comments on the test in case you touch that file again, but nothing blocking.

Not loading very old fee estimates makes good sense to me to further avoid creating stuck transactions.

ismaelsadeeq commented on test/functional/feature_fee_estimation.py:
```diff
@@ -290,6 +291,77 @@ def sanity_check_rbf_estimates(self, utxos):
+        fee_dat_current_content = open(fee_dat, "rb").read()
+        assert_equal(fee_dat_current_content, fee_dat_initial_content)
+
+        # Verify that the estimates are not the same if there are no blocks in the flush interval
```
Fixed Thank you.

instagibbs reviewed (approved):
code review ACK 6d36f3edbe7549cfaadc48b43c09dd3e581e92ff

I'd really prefer we have a regtest-only option to skip the check to impact integration tests the least.

ismaelsadeeq commented:
> I'd really prefer we have a regtest-only option to skip the check to impact integration tests the least.

I added a support for skipping the check only on regtest in 2fec7f362d8540cda5ff7e3e4dfb2cb751364e06 by passing -acceptstaleestimates option and it's tested in dea3accb1f2eeef74f293262c168d68c0ec444cb

glozow commented on src/policy/fees.h:
```diff
@@ -25,6 +25,9 @@
 // How often to flush fee estimates to fee_estimates.dat.
 static constexpr std::chrono::hours FEE_FLUSH_INTERVAL{1};
 
+// Stale fee estimates are not going to be read on mainnet
+static constexpr bool ACCEPT_STALE_FEE_ESTIMATES = false;
```
in 2fec7f362d8540cda5ff7e3e4dfb2cb751364e06:

prefix with `DEFAULT_`, make doxygen compatible, use braced initialization. Documentation about where the option can be used belongs where that code lives, not here.

```suggestion
/** Whether we allow importing a fee_estimates file older than MAX_FILE_AGE. */
static constexpr bool DEFAULT_ACCEPT_STALE_FEE_ESTIMATES{false};
```

ismaelsadeeq replied:
Thank you, I updated it.

glozow commented on src/init.cpp:
```diff
@@ -586,6 +586,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-acceptnonstdtxn", strprintf("Relay and mine \"non-standard\" transactions (%sdefault: %u)", "testnet/regtest only; ", !testnetChainParams->RequireStandard()), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::NODE_RELAY);
     argsman.AddArg("-incrementalrelayfee=<amt>", strprintf("Fee rate (in %s/kvB) used to define cost of relay, used for mempool limiting and replacement policy. (default: %s)", CURRENCY_UNIT, FormatMoney(DEFAULT_INCREMENTAL_RELAY_FEE)), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::NODE_RELAY);
     argsman.AddArg("-dustrelayfee=<amt>", strprintf("Fee rate (in %s/kvB) used to define dust, the value of an output such that it will cost more than its value in fees at this fee rate to spend it. (default: %s)", CURRENCY_UNIT, FormatMoney(DUST_RELAY_TX_FEE)), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::NODE_RELAY);
+    argsman.AddArg("-acceptstalefeeestimates", "Read fee estimates even if they are stale supported only in regtest", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```
in 2fec7f362d8540cda5ff7e3e4dfb2cb751364e06
Please also include here (1) the default value (2) how old "stale" is specifically, using the constants.

ismaelsadeeq replied:
I modify as you suggested 👍🏾

glozow commented on src/policy/fees.h:
```diff
@@ -192,9 +195,11 @@ class CBlockPolicyEstimator
      */
     static constexpr std::chrono::hours MAX_FILE_AGE{60};
 
+    bool m_read_stale_estimates;
```
2fec7f362d8540cda5ff7e3e4dfb2cb751364e06

this should be const

ismaelsadeeq replied:
Thank you

glozow commented on test/functional/feature_fee_estimation.py:
```diff
@@ -290,6 +292,95 @@ def sanity_check_rbf_estimates(self, utxos):
+
+    def test_estimate_dat_is_flushed_periodically(self):
+        fee_dat = self.nodes[0].chain_path / "fee_estimates.dat"
+        hour = 3600
```
nit in dea3accb1f2eeef74f293262c168d68c0ec444cb

maybe replace this `hour` and the one in the other tests with a constant at the top of the file, `SECONDS_PER_HOUR`

glozow commented on test/functional/feature_fee_estimation.py:
```diff
@@ -290,6 +292,95 @@ def sanity_check_rbf_estimates(self, utxos):
+        # Verify that fee estimates were flushed and fee_estimates.dat file is created
+        assert_equal(os.path.isfile(fee_dat), True)
+
+        # Verify that the estimates remain the same if there are no blocks in the flush interval
```
in dea3accb1f2eeef74f293262c168d68c0ec444cb

Given that you comment "there are no blocks" you should probably also assert that is true, e.g. `assert_equal` the result from `getbestblockhash` before and after

ismaelsadeeq replied:
sure fixed

glozow reviewed (commented):
Looks pretty good, thanks for adding the option.
Please clean up the commit messages. dea3accb1f2eeef74f293262c168d68c0ec444cb claims to add the regtest option, but that is done in 2fec7f362d8540cda5ff7e3e4dfb2cb751364e06.

ismaelsadeeq committed:
tx fees, policy: periodically flush fee estimates to fee_estimates.dat

This reduces chances of having old estimates in fee_estimates.dat.

instagibbs commented:
tsan failure looks unrelated: https://cirrus-ci.com/task/5970681108627456

instagibbs commented on src/init.cpp:
```diff
@@ -1253,7 +1255,18 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    if (!ignores_incoming_txs) {
+        bool read_stale_estimates = args.GetBoolArg("-acceptstalefeeestimates", DEFAULT_ACCEPT_STALE_FEE_ESTIMATES);
+        if (read_stale_estimates && (chainparams.GetChainType() != ChainType::REGTEST)) {
+            LogPrintf("Reading stale fee estimates is only supported in regtest \n");
```
I think this should fail rather than being ignored?

ismaelsadeeq replied:
I agree, fixed it.

willcl-ark commented on src/policy/fees.h:
```diff
@@ -14,12 +14,25 @@
+// How often to flush fee estimates to fee_estimates.dat.
+static constexpr std::chrono::hours FEE_FLUSH_INTERVAL{1};
+
+/** fee_estimates.dat that are more than 60 hours old, 2 and half days will not be read,
```
clarity nit:

```suggestion
/** fee_estimates.dat that are more than 60 hours (2.5 days) will not be read,
```

ismaelsadeeq replied:
Updated, thank you @willcl-ark

willcl-ark reviewed (approved):
re-ACK efe5f373e1

Changes look good. Left a nit not to be addressed unless you re-touch :)

I also checked how much time the new tests were adding to feature_fee_estimation.py, as this is often one of the longest-running tests on my machine. It looks like it's <=1second, so that seems fine.

pinheadmz reviewed (approved):
ACK efe5f373e12329104f1c706145eee75c2d2079a7

Great work on this! A few non-breaking questions below.
Ran all tests locally, ensured all 3 new tests fail on `master`.
Reviewed each commit, confirmed regtest-only argument is only accepted on regtest!

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK efe5f373e12329104f1c706145eee75c2d2079a7
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmSIvz0ACgkQ5+KYS2KJ
yTrL8w//coGfk9oOwTpePMIqxFQxlhxZP9hX6wur36aKPLtR7e07DYelDAvdQnZn
U3zLpMVDhoaWoAysQUinMyblQ06JCaNmPhDtu9b3gs6kPMGIhmhqM1SgcHtosvUe
r05sZGVZVlhz3lsHQ4ebXqzQwq6LuAxvXniKeTj7Afu2IhAsycJCFoXTSzTbMWlG
2GyJdkA1PQnaEMV5uZVFI7/oosG0ifNsoH9jeRYbqRoZjXQC3rAtP4a+T07IiNgv
K4lI1HrekW3M0+JpvoV38cYGuIehpKEoMS9ISTxa3o3zTNn8KjxpXZTOXq9o6IAv
5p7VpMHFVWXYBD+V21Jk2zni/BS6x2KtPLWhMNlTUoxnsw28R8pRbyeUEYBjIXEX
Z19BgwMOsZqWaz8DU0SedRC1ngpbidOrbgYQZaTXXP0BZX1YZDjTMH/ojV/gVWqo
VDhpT8RG/PetW6xvOAp9pAfDdNENvSCOIQVjCg04kxmqlrAvbNS9UHX+dvuCHEGB
njArT/p1TjCD3Pq6hO3YSH8ANVRnrb4Vg3qoB0WzT7DcHfySmLiydDnvmOut2HEC
G2D4vor14YXr5/NNipImX0IO0aTMXr5drswhwfXPoBjSExG4UQ6f4Lev2/AHkibH
5tSMfX40NagvSx9+xNqKjpZrKk+89rFhddfA9iB+XZidAzYSoVc=
=49Ly
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

ismaelsadeeq commented on test/functional/feature_fee_estimation.py:
```diff
@@ -290,6 +293,95 @@ def sanity_check_rbf_estimates(self, utxos):
+
+        # Stop the node and backdate the fee_estimates.dat file more than MAX_FILE_AGE
+        self.stop_node(0)
+        last_modified_time = time.time() - (MAX_FILE_AGE + 1) * SECONDS_PER_HOUR
```
61 hours.

achow101 commented:
@ismaelsadeeq Are you planning on making any further changes here?

ismaelsadeeq commented:
> @ismaelsadeeq Are you planning on making any further changes here?

Yes I will add all recent feedbacks in next push.

ismaelsadeeq committed:
tx fees, policy: do not read estimates of old fee_estimates.dat

Old fee estimates could cause transactions to become stuck in the
mempool. This commit prevents the node from using stale estimates
from an old file.

ismaelsadeeq committed:
tx fees, policy: read stale fee estimates with a regtest-only option

If -acceptstalefeeestimates option is passed stale fee estimates can now
be read when operating in regtest environments.

Additionally, this commit updates all declarations of the CBlockPolicyEstimator
class to include a the second constructor variable.

ismaelsadeeq committed:
test: ensure old fee_estimate.dat not read on restart and flushed

This commit adds tests to ensure that old fee_estimates.dat files
are not read and that fee_estimates are periodically flushed to the
fee_estimates.dat file.

Additionaly it tests the -regtestonly option -acceptstalefeeestimates.

ismaelsadeeq commented on src/policy/fees.cpp:
```diff
@@ -528,8 +528,9 @@ bool CBlockPolicyEstimator::_removeTx(const uint256& hash, bool inBlock)
 
-CBlockPolicyEstimator::CBlockPolicyEstimator(const fs::path& estimation_filepath)
-    : m_estimation_filepath{estimation_filepath}
+CBlockPolicyEstimator::CBlockPolicyEstimator(const fs::path& estimation_filepath, const bool& read_stale_estimates)
```
Yes, thank you. I updated it

ismaelsadeeq commented on src/policy/fees.h:
```diff
@@ -191,9 +194,12 @@ class CBlockPolicyEstimator
 
     const fs::path m_estimation_filepath;
+
+    const bool m_read_stale_estimates;
```
No, It does not thank you.

ismaelsadeeq commented on src/policy/fees.h:
```diff
@@ -191,9 +194,12 @@ class CBlockPolicyEstimator
 public:
     /** Create new BlockPolicyEstimator and initialize stats tracking classes with default values */
-    CBlockPolicyEstimator(const fs::path& estimation_filepath);
+    CBlockPolicyEstimator(const fs::path& estimation_filepath, const bool& read_stale_estimates);
```
Updated 💯

instagibbs commented:
@ismaelsadeeq just FYI if you don't have to rebase on master due to a conflict, don't. It's extra review burden and makes getting re-ACKs harder.

ismaelsadeeq commented:
> @ismaelsadeeq just FYI if you don't have to rebase on master due to a conflict, don't. It's extra review burden and makes getting re-ACKs harder.

Ohh Sorry, Thank you for pointing this out.
Its okay now I think and any further nit changes I would address in a follow-up.

willcl-ark commented:
Just noting my understanding of this and the followup [pull](https://github.com/bitcoin/bitcoin/pull/27859) and how they work together.

In this PR we will regularly persist fee estimations to _fee_estimates.dat_, providing more protection in an unclean shutdown scenario. We then reload these on startup, under then new condition that they are less than 2.5 days old.

1. In the case where the restart is instant/very fast this let's us keep accurate fee estimation after a crash.
2. In the case where the restart is > 60 hours then we are improved in that we don't serve known stale data.
3. In the case where fees start high (or low), and remain high (or low), this will help us with more accurate fee estimation after restart.
4. If fees start high and descrease then we may have too-high fee estimation, which still helps us avoid unwated stuck transactions.
5. If fees start low and increased then our fee estimation may report too-low. Although this pull doesn't help directly in this case, it's not worse than current behaviour and the expiry should make it strictly better.

In the followup pull, persisting mempool_min_fee to disk is much more sensitive than fee estimation data. If we used outdated rates here we may start rejecting valid tx from the p2p network from our mempool which may be undesirable. Therefore IMO it makes sense there to be much more strict about the lifetime of mempool_min_fee rates read from disk at startup; perhaps something more in the order of 1 hour?

RandyMcMillan commented:
In this scenario - the node is going to have to sync - whether it was shut down for 1 hour or 60+ hours. IMO it would be ideal to not have **one more thing** dependent upon system time when booting up the node (*There are other issues related to system time when booting up the node). Since we know that the node must connect and sync - I think the ideal is to delay the file age check later into the life cycle of the node.

*In the scenario where system time has been skewed - off by many hours - there is a possibility that the file age is "in the future". You may want to use absolute value in the file age check... this will remove the possibility of a negative value for file age.

instagibbs commented:
> Since we know that the node must connect and sync - I think the ideal is to delay the file age check later into the life cycle of the node.

Not sure what the suggestion is. Can you be specific?

willcl-ark commented on test/functional/feature_fee_estimation.py:
```diff
@@ -290,6 +293,95 @@ def sanity_check_rbf_estimates(self, utxos):
+        assert fee_dat_current_content != fee_dat_initial_content
+
+
+    def test_acceptstalefeeestimates_option(self):
```
nit: There's no test to ensure that stale estimates _aren't_ accepted on non-regtest networks, and I'm not sure there particularly needs to be. But I verified manually that this was the case:

```log
$./src/bitcoind -testnet -acceptstalefeeestimates
2023-06-20T12:47:07Z Error: acceptstalefeeestimates is not supported on test chain.
2023-06-20T12:47:07Z Shutdown: In progress...
2023-06-20T12:47:07Z scheduler thread exit
2023-06-20T12:47:07Z Shutdown: done
```

ismaelsadeeq replied:
Fixed in #28157

willcl-ark reviewed (approved):
ACK d2b39e09bc

Looks good to me now.

Using stale fee estimates is a good way to get your transaction stuck in the mempool, and therefore not using those makes sense. I also agree with flushing the data periodically and not just on shutdown, which can help us have access to current estimates even after an unclean shutdown.

Left a few nits which don't need addressing

instagibbs commented on src/policy/fees.cpp:
```diff
@@ -545,9 +546,22 @@ CBlockPolicyEstimator::CBlockPolicyEstimator(const fs::path& estimation_filepath
     AutoFile est_file{fsbridge::fopen(m_estimation_filepath, "rb")};
-    if (est_file.IsNull() || !Read(est_file)) {
+
+    // Whenever the fee estimation file is not present return early
```
Not blocking, but I agree. Comments are generally required when the code fails to self-document. This whole section is self-explanatory, which is good, meaning comments aren't really required.

ismaelsadeeq replied:
Fixed in #28157

instagibbs reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/27622/commits/d2b39e09bc6a5982fc5cf4b538b7fdb0e3cae576

glozow reviewed (commented):
ACK d2b39e09bc6a5982fc5cf4b538b7fdb0e3cae576. One nit if you follow up.

dergoegge commented:
This should probably be backported.

ismaelsadeeq commented on src/policy/fees.h:
```diff
@@ -25,6 +25,11 @@
 // How often to flush fee estimates to fee_estimates.dat.
 static constexpr std::chrono::hours FEE_FLUSH_INTERVAL{1};
 
+/** fee_estimates.dat that are more than 60 hours (2.5 days) will not be read,
+ * as the estimates in the file are stale.
```
Fixed in #28157

fanquake commented:
Added to #28487 for backporting to 25.1.

fanquake commented:
Pushed into #28535 for backporting to 24.x as well.
